homeassistant/helpers/state.py

```diff
@@ -132,6 +99,103 @@ def get_changed_since(states: Iterable[State],
             if state.last_updated >= utc_point_in_time]
 
 
+@bind_hass
+def async_map_state_to_calls(hass: HomeAssistantType,
+                             domain: str,
+                             state: State,
+                             state_to_service: Dict[str, str],
+                             services: Dict[str,
+                                            voluptuous.schema_builder.Schema]
+                             ) -> List[ServiceCall]:
+        """
+        Maps a state object into service calls to reproduce the given state.
+
+        Function will make two passes. First it will create a call to reproduce
+        the state.state, passing along any allowed attributes to the state
+        setting service. As a second pass it will map remaining attributes
+        to services
+        """
+
+        calls = []
+
+        # expanded attribute list containing entity id
+        expanded_attributes = dict(state.attributes)
+        expanded_attributes[ATTR_ENTITY_ID] = state.entity_id
+
+        def add_if_complete(service: str) -> None:
+            """Add a service call with the attributes needed."""
+            schema = voluptuous.schema_builder.Schema(
+                services[service].schema,
+                required=services[service].required,
+                extra=voluptuous.schema_builder.REMOVE_EXTRA)
+
+            try:
+                attributes = schema(expanded_attributes)
+            except voluptuous.error.Invalid as e:
+                _LOGGER.debug("Incomplete attribute "
+                              "list to call {}:{}".format(service, e))
+                return
+
+            # consume attributes taken by service
+            # this avoid multiple service calls
+            # for same attribute if schema would
+            # allow it
+            for key in attributes:
+                expanded_attributes.pop(key, None)
+
+            calls.append(
+                ServiceCall(domain,
+                            service,
+                            data=attributes,
+                            context=state.context))
+
+        # first try to reproce the state value, taking all attributes
+        if state.state in state_to_service:
```

Comment:
The following code block might also be better to keep in each component. The priority between states and state attributes might be different for different components.

Reply:
sure, i'll split to two calls.